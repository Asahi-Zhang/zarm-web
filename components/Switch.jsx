
import React, { Component, PropTypes, Children } from 'react';
import classnames from 'classnames';

class Switch extends Component {

  constructor(props) {
    super(props);
    this.state = {
      checked: false
    };
  }

  componentDidMount() {
    const props = this.props;
    let checked = false;
    if ('checked' in props) {
      checked = !!props.checked;
    } else {
      checked = !!props.defaultChecked;
    }
    this.setState({
      checked: checked,
    });
  }

  componentWillReceiveProps(nextProps) {
    if ('checked' in nextProps) {
      this.setState({
        checked: !!nextProps.checked
      });
    }
  }

  _onClick() {
    const checked = !this.state.checked;
    this.setState({
      checked: checked
    });
    this.props.onChange(checked);
  }

  render() {
    const { size, isCheckedText, unCheckedText, ...others } = this.props;
    const { checked } = this.state;

    const cls = classnames({
      'ui-switch'        : true,
      ['size-' + size]   : size,
      'ui-switch-checked': checked,
    });

    return (
      <div className={cls} onClick={() => this._onClick()} {...others}>
        <span className="ui-switch-inner">{ checked ? isCheckedText : unCheckedText }</span>
      </div>
    );
  }
}

Switch.propTypes = {
  size         : PropTypes.oneOf(['sm']),
  defaultValue : PropTypes.bool,
  isCheckedText: PropTypes.string,
  unCheckedText: PropTypes.string,
  onChange     : PropTypes.func,
};

Switch.defaultProps = {
  defaultValue : false,
  isCheckedText: '',
  unCheckedText: '',
  onChange     : function () {},
};

export default Switch;