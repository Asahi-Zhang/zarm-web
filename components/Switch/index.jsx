
import React, { Component, PropTypes, Children } from 'react';
import classnames from 'classnames';

class Switch extends Component {

  constructor(props) {
    super(props);
    this.state = {
      value: props.value || props.defaultValue,
    };
  }

  componentWillReceiveProps(nextProps) {
    if ('value' in nextProps) {
      this.setState({
        value: !!nextProps.value
      });
    }
  }

  render() {
    const { size, isCheckedText, unCheckedText, ...others } = this.props;
    const { value } = this.state;

    const cls = classnames({
      'ui-switch'        : true,
      ['size-' + size]   : size,
      'ui-switch-checked': value,
    });

    return (
      <span className={cls} onClick={() => this._onClick()} {...others}>
        <span className="ui-switch-inner">{ value ? isCheckedText : unCheckedText }</span>
      </span>
    );
  }

  _onClick() {
    const value = !this.state.value;
    this.setState({
      value: value
    });
    this.props.onChange(value);
  }
}

Switch.propTypes = {
  size         : PropTypes.oneOf(['sm']),
  value        : PropTypes.bool,
  defaultValue : PropTypes.bool,
  isCheckedText: PropTypes.oneOfType([PropTypes.string, PropTypes.element]),
  unCheckedText: PropTypes.oneOfType([PropTypes.string, PropTypes.element]),
  onChange     : PropTypes.func,
};

Switch.defaultProps = {
  size         : null,
  defaultValue : false,
  isCheckedText: '',
  unCheckedText: '',
  onChange     : function () {},
};

export default Switch;