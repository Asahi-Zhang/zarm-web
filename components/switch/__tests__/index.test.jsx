import React from 'react';
import { render, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import Switch from '../index';

describe('Switch', () => {
  it('renders normal Switch correctly', () => {
    const wrapper = render(
      <div>
        <Switch />
      </div>,
    );

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders Switch correctly with more props', () => {
    const wrapper = render(
      <div>
        <Switch disabled size="sm" />
      </div>,
    );

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('behaves correctly when receving new value', () => {
    const wrapper = mount(
      <Switch defaultChecked={false} />,
    );

    expect(wrapper.find('.zw-switch').hasClass('zw-switch--checked')).toBeFalsy();
    wrapper.find('.zw-switch').simulate('click');
    expect(wrapper.find('.zw-switch').hasClass('zw-switch--checked')).toBeTruthy();
  });

  it('behaves correctly when toggling status', () => {
    const onChange = jest.fn();
    const wrapper = mount(
      <Switch defaultChecked={false} onChange={onChange} />,
    );

    const wrapperOpen = mount(
      <Switch defaultChecked={false} onChange={onChange} />,
    );

    wrapper.find('.zw-switch').simulate('click');
    expect(onChange).toHaveBeenCalledWith(true);

    wrapperOpen.find('.zw-switch').simulate('click');
    expect(onChange).toHaveBeenCalledWith(true);
  });

  it('switch with loading status', () => {
    const wrapper = mount(
      <Switch loading defaultChecked={true} />,
    );
    expect(wrapper.find('.zw-switch__loading').length > 0).toBeTruthy();

    const wrapperOpen = mount(
      <Switch defaultChecked={true} loading />,
    );
    expect(wrapperOpen.find('.zw-switch__loading').length > 0).toBeTruthy();

  });
});
