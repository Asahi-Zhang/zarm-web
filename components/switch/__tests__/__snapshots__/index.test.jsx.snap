// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch renders Switch correctly with more props 1`] = `
<div>
  <span
    class="zw-switch--disabled zw-switch--sm"
  >
    <span
      class="zw-switch__inner"
    >
      <span />
    </span>
  </span>
</div>
`;

exports[`Switch renders normal Switch correctly 1`] = `
<div>
  <span
    class="zw-switch"
  >
    <span
      class="za-switch__inner"
    >
      <span />
    </span>
  </span>
</div>
`;
